File: src/api/federation/proxy.py
import base64
import ipaddress
import urllib.parse

import httpx
from fastapi import Request
from starlette.background import BackgroundTask
from starlette.responses import StreamingResponse

from api.errors import ErrorCodes, NexusGateException
from config.provider import GlobalConfigProvider


def get_proxy_client(request: Request, verify_ssl: bool = True) -> httpx.AsyncClient:
    """Manages globally persistent proxy connections mapping trust states attached to the ASGI app."""
    if not hasattr(request.app.state, "http_clients"):
        request.app.state.http_clients = {}

    clients = request.app.state.http_clients
    if verify_ssl not in clients:
        limits = httpx.Limits(max_connections=50, max_keepalive_connections=10)
        clients[verify_ssl] = httpx.AsyncClient(
            limits=limits, timeout=httpx.Timeout(30.0, connect=5.0), verify=verify_ssl
        )
    return clients[verify_ssl]


# ─────────────────────────────────────────────────────────────────────────────
# Alias Resolution
# ─────────────────────────────────────────────────────────────────────────────

_ALIAS_TO_SERVER: dict = {}


def _build_alias_map():
    """Builds O(1) alias to server routing table based on configured federation alias mappings."""
    global _ALIAS_TO_SERVER
    _ALIAS_TO_SERVER.clear()

    config = GlobalConfigProvider().get_config()

    # Pre-populate explicitly mapped federated aliases from database configs
    for db_alias, db_config in config.database.items():
        fed_alias = getattr(db_config, "federated_alias", None)
        if isinstance(fed_alias, str):
            for srv_alias in config.federation.server:
                if fed_alias.startswith(f"{srv_alias}_"):
                    _ALIAS_TO_SERVER[fed_alias] = srv_alias
                    break

    # Pre-populate implicit ones that might be accessed directly without a local db config
    # In a full overhaul, this would sync with federation state
    config.federation.alias_map = _ALIAS_TO_SERVER


def _resolve_server(alias: str) -> str | None:
    """O(1) resolution from full alias to target server ID."""
    if not _ALIAS_TO_SERVER:
        _build_alias_map()

    server_id = _ALIAS_TO_SERVER.get(alias)
    if server_id:
        return server_id

    # Fallback for dynamic aliases not pre-mapped
    config = GlobalConfigProvider().get_config()
    for srv_alias in config.federation.server:
        if alias.startswith(f"{srv_alias}_"):
            _ALIAS_TO_SERVER[alias] = srv_alias
            return srv_alias

    return None


# ─────────────────────────────────────────────────────────────────────────────
# Proxy Request Builders
# ─────────────────────────────────────────────────────────────────────────────


def _build_remote_url(
    srv_config, target_alias: str, path: str, query: str, is_database: bool
) -> str:
    """Generates the absolute upstream endpoint bypassing string concats."""
    base_url = srv_config.url.rstrip("/")
    subpath = path.lstrip("/")

    remote_url = (
        f"{base_url}/api/db/{target_alias}/{subpath}"
        if is_database
        else f"{base_url}/api/fs/{target_alias}/{subpath}"
    )
    return f"{remote_url}?{query}" if query else remote_url


def _build_proxy_headers(request: Request, srv_config) -> dict:
    """Constructs transmission limits erasing native Host overlays."""
    headers = dict(request.headers)
    headers.pop("host", None)
    headers.pop("content-length", None)

    encoded_secret = base64.b64encode(srv_config.secret.encode("utf-8")).decode("utf-8")
    headers["X-Federation-Secret"] = encoded_secret
    headers["X-Federation-Node"] = srv_config.node_id
    headers["X-Request-ID"] = getattr(request.state, "request_id", "-")

    return headers


async def _stream_proxy_execution(
    client: httpx.AsyncClient, request: Request, remote_url: str, headers: dict
) -> StreamingResponse:
    """Dispatches background streaming sockets returning bound payloads."""
    req = client.build_request(
        request.method,
        remote_url,
        headers=headers,
        content=request.stream()
        if request.method in ("POST", "PUT", "PATCH")
        else None,
    )

    try:
        resp = await client.send(req, stream=True)
        pass_headers = {
            k.lower(): v
            for k, v in resp.headers.items()
            if k.lower()
            not in (
                "transfer-encoding",
                "content-encoding",
                "connection",
                "content-length",
            )
        }

        return StreamingResponse(
            resp.aiter_bytes(),
            status_code=resp.status_code,
            headers=pass_headers,
            background=BackgroundTask(resp.aclose),
        )
    except httpx.RequestError as req_error:
        raise NexusGateException(
            ErrorCodes.FED_SERVER_DOWN, f"Federated server error: {str(req_error)}", 502
        )


# ─────────────────────────────────────────────────────────────────────────────
# Primary Dispatcher
# ─────────────────────────────────────────────────────────────────────────────


def _is_safe_url(url: str) -> bool:
    """Blocks SSRF loops and strictly protects internal bogon subnet spaces."""
    try:
        parsed = urllib.parse.urlparse(url)
        hostname = parsed.hostname
        if not hostname:
            return False

        if hostname.lower() in ("localhost", "127.0.0.1", "0.0.0.0", "::1"):
            return False

        try:
            ip = ipaddress.ip_address(hostname)
            if ip.is_private or ip.is_loopback or ip.is_link_local:
                return False
        except ValueError:
            pass  # Domain name, DNS resolution handles standard external routing

        return True
    except Exception:
        return False


async def proxy_request(
    alias: str, path: str, request: Request, is_database: bool = True
) -> StreamingResponse:
    """Entrypoint binding exact aliases targeting mapped proxies natively."""
    config = GlobalConfigProvider().get_config()

    srv_alias = _resolve_server(alias)

    if srv_alias and srv_alias in config.federation.server:
        srv_config = config.federation.server[srv_alias]
        target_alias = alias[len(srv_alias) + 1 :]
        remote_url = _build_remote_url(
            srv_config, target_alias, path, request.url.query, is_database
        )

        if not _is_safe_url(remote_url):
            raise NexusGateException(
                ErrorCodes.FED_SERVER_DOWN,
                "SSRF Blocked: Federation target resolves to an internal or restricted network.",
                403,
            )

        headers = _build_proxy_headers(request, srv_config)

        client = get_proxy_client(request, srv_config.trust_mode == "verify")
        return await _stream_proxy_execution(client, request, remote_url, headers)

    resource_type = "Database" if is_database else "Storage"
    raise NexusGateException(
        ErrorCodes.FED_SERVER_DOWN,
        f"Federated {resource_type} alias '{alias}' not found or unreachable",
        404,
    )

File: src/api/federation/proxy.rs
use axum::http::StatusCode;
use std::net::IpAddr;
use url::Url;

use crate::api::errors::AxiomError;
use crate::config::loader::ConfigManager;

/// SSRF guard — blocks private/loopback/link-local addresses
pub fn is_safe_url(raw_url: &str) -> bool {
    let parsed = match Url::parse(raw_url) {
        Ok(u) => u,
        Err(_) => return false,
    };

    let host = match parsed.host_str() {
        Some(h) => h,
        None => return false,
    };

    match host.to_lowercase().as_str() {
        "localhost" | "127.0.0.1" | "0.0.0.0" | "::1" => return false,
        _ => {}
    }

    // If the host is a raw IP, check it isn't private
    if let Ok(ip) = host.parse::<IpAddr>() {
        if ip.is_loopback() || ip.is_multicast() {
            return false;
        }
        match ip {
            IpAddr::V4(v4) => {
                if v4.is_private() || v4.is_link_local() || v4.is_broadcast() {
                    return false;
                }
            }
            IpAddr::V6(v6) => {
                if v6.is_loopback() {
                    return false;
                }
            }
        }
    }

    true
}

/// Resolve a federated alias (e.g. "node1_mydb") to the server alias ("node1")
/// and return the target sub-alias ("mydb").
pub fn resolve_alias(full_alias: &str) -> Option<(String, String)> {
    let config = ConfigManager::get();
    let fed = &config.federation;

    for srv_alias in fed.server.keys() {
        let prefix = format!("{}_", srv_alias);
        if full_alias.starts_with(&prefix) {
            let target = full_alias[prefix.len()..].to_string();
            return Some((srv_alias.to_string(), target));
        }
    }
    None
}

/// Build federation auth headers for an outgoing request to a peer node
pub fn build_fed_headers(node_id: &str, secret: &str) -> reqwest::header::HeaderMap {
    let mut headers = reqwest::header::HeaderMap::new();
    use base64::Engine;
    let b64 = base64::engine::general_purpose::STANDARD.encode(secret.as_bytes());
    if let (Ok(sec), Ok(nid)) = (
        reqwest::header::HeaderValue::from_str(&b64),
        reqwest::header::HeaderValue::from_str(node_id),
    ) {
        headers.insert("X-Federation-Secret", sec);
        headers.insert("X-Federation-Node", nid);
    }
    headers
}

/// Stream-proxy a request to the upstream federated server.
/// Returns the body bytes and upstream status code.
pub async fn proxy_request(
    full_alias: &str,
    path: &str,
    query: &str,
    method: &str,
    body: Option<bytes::Bytes>,
    is_database: bool,
    original_headers: &reqwest::header::HeaderMap,
) -> Result<(u16, bytes::Bytes, String), AxiomError> {
    let config = ConfigManager::get();

    let (srv_alias, target_alias) = resolve_alias(full_alias)
        .ok_or_else(|| AxiomError::new("FED_ALIAS_NOT_FOUND", "Federated alias not found", StatusCode::NOT_FOUND))?;

    let fed = &config.federation;

    let srv_cfg = fed.server.get(&srv_alias)
        .ok_or_else(|| AxiomError::new("FED_SERVER_NOT_FOUND", "Federation server config missing", StatusCode::NOT_FOUND))?;

    let base = srv_cfg.url.trim_end_matches('/');
    let subpath = path.trim_start_matches('/');
    let resource = if is_database { "db" } else { "fs" };
    let remote_url = if query.is_empty() {
        format!("{}/api/v1/{}/{}/{}", base, resource, target_alias, subpath)
    } else {
        format!("{}/api/v1/{}/{}/{}?{}", base, resource, target_alias, subpath, query)
    };

    if !is_safe_url(&remote_url) {
        return Err(AxiomError::new(
            "FED_SSRF_BLOCKED",
            "SSRF blocked: federation target resolves to an internal or restricted network",
            StatusCode::FORBIDDEN,
        ));
    }

    let mut fed_headers = build_fed_headers(&srv_cfg.node_id, &srv_cfg.secret);
    // Forward original auth header
    if let Some(auth) = original_headers.get("authorization") {
        fed_headers.insert("authorization", auth.clone());
    }

    let verify_ssl = srv_cfg.trust_mode == "verify";
    let client = reqwest::Client::builder()
        .danger_accept_invalid_certs(!verify_ssl)
        .timeout(std::time::Duration::from_secs(30))
        .default_headers(fed_headers)
        .build()
        .map_err(|e| AxiomError::new("FED_CLIENT_ERROR", &e.to_string(), StatusCode::INTERNAL_SERVER_ERROR))?;

    let req = match method.to_uppercase().as_str() {
        "POST" => client.post(&remote_url).body(body.unwrap_or_default()),
        "PUT" => client.put(&remote_url).body(body.unwrap_or_default()),
        "DELETE" => client.delete(&remote_url),
        _ => client.get(&remote_url),
    };

    let resp = req.send().await
        .map_err(|e| AxiomError::new("FED_SERVER_DOWN", &e.to_string(), StatusCode::BAD_GATEWAY))?;

    let status = resp.status().as_u16();
    let content_type = resp.headers()
        .get(reqwest::header::CONTENT_TYPE)
        .and_then(|v| v.to_str().ok())
        .unwrap_or("application/json")
        .to_string();

    let body_bytes = resp.bytes().await
        .map_err(|e| AxiomError::new("FED_READ_ERROR", &e.to_string(), StatusCode::BAD_GATEWAY))?;

    Ok((status, body_bytes, content_type))
}
